js/private/js_run_devserver.mjs

```diff
@@ -54,42 +53,128 @@ function isNodeModulePath(srcPath) {
+        })
+        output.once('readable', () => {
+            resolve(output.read().toString('hex'))
+        })
+        input.pipe(output)
+    })
+}
+
+// Converts a size in bytes to a human readable friendly number such as "24 KiB"
+export function friendlyFileSize(bytes) {
+    if (!bytes) {
+        return '0 B'
+    }
+    const e = Math.floor(Math.log(bytes) / Math.log(1024))
+    if (e == 0) {
+        return `${bytes} B`
+    }
+    return (
+        (bytes / Math.pow(1024, Math.min(e, 5))).toFixed(1) +
+        ' ' +
+        ' KMGTP'.charAt(Math.min(e, 5)) +
+        'iB'
+    )
+}
+
+// https://stackoverflow.com/a/42299191
+function partitionArray(array, callback) {
+    return array.reduce(
+        function (result, element, i) {
+            callback(element, i, array)
+                ? result[0].push(element)
+                : result[1].push(element)
+
+            return result
+        },
+        [[], []]
+    )
+}
+
 // Recursively copies a file, symlink or directory to a destination. If the file has been previously
 // synced it is only re-copied if the file's last modified time has changed since the last time that
 // file was copied. Symlinks are not copied but instead a symlink is created under the destination
 // pointing to the source symlink.
-async function syncRecursive(src, dst, writePerm) {
+async function syncRecursive(src, dst, sandbox, writePerm) {
     try {
         const lstat = await fs.promises.lstat(src)
-        const last = synced.get(src)
+        const last = syncedTime.get(src)
         if (!lstat.isDirectory() && last && lstat.mtimeMs == last) {
             // this file is already up-to-date
+            if (process.env.JS_BINARY__LOG_DEBUG) {
+                console.error(
+                    `Skipping file ${src.slice(
+                        RUNFILES_ROOT.length + 1
+                    )} since its timestamp has not changed`
+                )
+            }
             return 0
         }
-        const exists = synced.has(src) || fs.existsSync(dst)
-        synced.set(src, lstat.mtimeMs)
+        const exists = syncedTime.has(src) || fs.existsSync(dst)
+        syncedTime.set(src, lstat.mtimeMs)
         if (lstat.isSymbolicLink()) {
             const srcWorkspacePath = src.slice(RUNFILES_ROOT.length + 1)
-            if (process.env.JS_BINARY__LOG_DEBUG) {
-                console.error(`Syncing symlink ${srcWorkspacePath}`)
-            }
+            let symlinkMeta = ''
             if (isNodeModulePath(src)) {
-                // Special case for node_modules symlinks where we should _not_ symlink to the runfiles but rather
-                // the bin copy of the symlink to avoid finding npm packages in multiple node_modules trees
-                const maybeBinSrc = path.join(
-                    process.env.JS_BINARY__EXECROOT,
-                    process.env.JS_BINARY__BINDIR,
-                    srcWorkspacePath
+                let linkPath = fs.readlinkSync(src)
```

Comment:
We're in an `async` method so this could be `await fs.promises.readLink(srcs)`? Or is this one of the few methods with no async version?